### PR TITLE
Add deprecation handling in backend

### PR DESCRIPTION
**Pull request by klecki**
Keeps the deprecation warnings in Python.
Renaming and detecting double arguments handled in backend.
Skipping unused arguments done in backend.

Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>

#### Why we need this PR?
- Refactoring to handle deprecations for all apis at once.

#### What happened in this PR? 
 - What solution was applied:
    Keeps the deprecation warnings in Python.
    Renaming and detecting double arguments handled in backend.
    Skipping unused arguments done in backend.

 - Affected modules and functionalities:
     OpSpec, ops.py
 - Key points relevant for the review:
	 Is this enough? Do we want warnings in C++?
 - Validation and testing:
     Test added. Can we add dummy operator for testing in Python?
     Should we cover serialization somehow? It requires some fake serialized data + deserialization + more complete, nontrivial pipeline for test. (probably buildable one to allow for verification).
 - Documentation (including examples): N/A

**JIRA TASK**: *[Use DALI-XXXX or NA]*

### CODE REVIEW

**Comment by klecki** [comment]
!build

File: dali/python/nvidia/dali/ops.py [hunk 1]
```diff
@@ -380,17 +380,6 @@ def __init__(self, **kwargs):
                 if not self._schema.IsDeprecatedArg(arg_name):
                     continue
                 meta = self._schema.DeprecatedArgMeta(arg_name)
-                new_name = meta['renamed_to']
```
Comment (mzient) [inline dali/python/nvidia/dali/ops.py:383]: Do I understand it correctly that now Python only issues warnings, but the errors are left to the backend?

**Comment by klecki** [comment]
!build

File: dali/python/nvidia/dali/ops.py (hunk 1 above)
Comment (klecki) [reply dali/python/nvidia/dali/ops.py:383]: Yes

File: dali/pipeline/operator/op_spec.h [hunk 2]
```diff
@@ -94,6 +94,7 @@ class DLL_PUBLIC OpSpec {
    */
   template <typename T>
   DLL_PUBLIC inline OpSpec& AddArg(const string &name, const T &val) {
+    VerifyRenamedAfterDeprecated(name);
```
Comment (JanuszL) [inline dali/pipeline/operator/op_spec.h:97]: Why here?
Wont be called anyway in the chain of AddArg->SetArg->SetInitializedArg->VerifyRenamedAfterDeprecated?

Comment (klecki) [reply dali/pipeline/operator/op_spec.h:97]: We need to issue this error before the regular error of duplicated arguments. So I added it in all places where we have the check for duplicated arguments.

**Review by JanuszL — APPROVED** [review]

**Comment by klecki** [comment]
!build

**Review by JanuszL — APPROVED** [review]

File: dali/pipeline/operator/op_spec.h [hunk 3]
```diff
@@ -135,20 +136,56 @@ class DLL_PUBLIC OpSpec {
...
         "\" already exists. ");
-    arguments_[name].reset(arg);
-    return *this;
+    return SetInitializedArg(name, arg);
   }
 
   /**
    * @brief Sets or adds an argument with given name
+   *
+   * Routes the deprecated argument to appropriate name or drops them
```
Comment (mzient) [inline dali/pipeline/operator/op_spec.h:149]:
```suggestion
   * @remarks Deprecated arguments are renamed (or dropped, if no longer used).
```

File: dali/pipeline/operator/op_spec.h [hunk 4]
```diff
@@ -135,20 +136,56 @@ class DLL_PUBLIC OpSpec {
...
    */
-  DLL_PUBLIC inline OpSpec& SetInitializedArg(const string& name, Argument* arg) {
-    arguments_[name].reset(arg);
+  DLL_PUBLIC inline OpSpec& SetInitializedArg(const string& arg_name, Argument* arg) {
+    if (schema_ && schema_->IsDeprecatedArg(arg_name)) {
+      const auto& deprecation_meta = schema_->DeprecatedArgMeta(arg_name);
+      // Argument is removed, and we can discard it
```
Comment (mzient) [inline dali/pipeline/operator/op_spec.h:154]:
```suggestion
      // Argument was removed, and we can discard it
```

File: dali/pipeline/operator/op_spec.h [hunk 5]
```diff
@@ -135,20 +136,56 @@ class DLL_PUBLIC OpSpec {
...
+    }
+    VerifyRenamedAfterDeprecated(arg_name);
+    arguments_[arg_name].reset(arg);
     return *this;
   }
+
+  /**
+   * @brief Check if the `arg_name` was already set through deprecated argument
+   */
+  DLL_PUBLIC inline void VerifyRenamedAfterDeprecated(const string& arg_name) {
```
Comment (mzient) [inline dali/pipeline/operator/op_spec.h:182]: verify = check if something is true / authentic / real.

```suggestion
  DLL_PUBLIC inline void CheckIfSetWithAnotherName(const string& arg_name) {
```
or
```suggestion
  DLL_PUBLIC inline void EnforceUniqueArgName(const string& arg_name) {
```

File: dali/pipeline/operator/op_spec.h [hunk 6]
```diff
@@ -455,6 +492,9 @@ class DLL_PUBLIC OpSpec {
   std::unordered_map<string, std::shared_ptr<Argument>> arguments_;
   std::unordered_map<string, Index> argument_inputs_;
   std::set<Index> argument_inputs_indexes_;
+  // Regular arguments that were already set through renamed deprecated arguments
+  // Maps regular_argument -> deprecated_argument
+  std::map<std::string, std::string> set_through_deprecated_arguments_;
```
Comment (mzient) [inline dali/pipeline/operator/op_spec.h:497]:
```suggestion
  std::unordered_map<std::string, std::string> set_through_deprecated_arguments_;
```

File: dali/pipeline/operator/op_spec_test.cc [hunk 7]
```diff
@@ -190,6 +193,34 @@ TEST(OpSpecTest, GetArgumentNonExisting) {
...
+  ASSERT_EQ(spec0.GetArgument<int>("replacing_arg"), 1);
+  int result = 0;
+  ASSERT_FALSE(spec0.TryGetArgument<int>(result, "deprecated_arg"));
+  ASSERT_TRUE(spec0.TryGetArgument<int>(result, "replacing_arg"));
+  ASSERT_EQ(result, 1);
+
+  ASSERT_THROW(OpSpec("DummyOpForSpecTest")
+      .AddArg("batch_size", 2)
+      .AddArg("deprecated_arg", 1)
+      .AddArg("replacing_arg", 2), std::runtime_error);
```
Comment (mzient) [inline dali/pipeline/operator/op_spec_test.cc:210]:
```suggestion
      .AddArg("replacing_arg", 2), DALIException);
```

File: dali/pipeline/operator/op_spec.h (hunk 6 above)
Comment (mzient) [reply dali/pipeline/operator/op_spec.h:497]: OK

File: dali/pipeline/operator/op_spec.h (hunk 3 above)
Comment (klecki) [reply dali/pipeline/operator/op_spec.h:149]: done

File: dali/pipeline/operator/op_spec.h (hunk 4 above)
Comment (klecki) [reply dali/pipeline/operator/op_spec.h:154]: done

File: dali/pipeline/operator/op_spec.h [hunk 8]
```diff
@@ -135,20 +136,56 @@ class DLL_PUBLIC OpSpec {
...
+        return *this;
+      }
+    }
+    VerifyRenamedAfterDeprecated(arg_name);
+    arguments_[arg_name].reset(arg);
     return *this;
   }
+
+  /**
+   * @brief Check if the `arg_name` was already set through deprecated argument
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/pipeline/operator/op_spec.h:180]: Done

File: dali/pipeline/operator/op_spec.h (hunk 5 above)
Comment (klecki) [reply dali/pipeline/operator/op_spec.h:182]: Done

File: dali/pipeline/operator/op_spec_test.cc [hunk 9]
```diff
@@ -190,6 +193,34 @@ TEST(OpSpecTest, GetArgumentNonExisting) {
...
+
+  ASSERT_THROW(OpSpec("DummyOpForSpecTest")
+      .AddArg("batch_size", 2)
+      .AddArg("replacing_arg", 1)
+      .AddArg("deprecated_arg", 2), std::runtime_error);
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/pipeline/operator/op_spec_test.cc:215]: done

File: dali/pipeline/operator/op_spec_test.cc (hunk 7 above)
Comment (klecki) [reply dali/pipeline/operator/op_spec_test.cc:210]: done

**Comment by klecki** [comment]
!build

**Review by mzient — APPROVED** [review]

**Pull request merged**